cpp/src/plasma/tf_plasma_op.cc

```diff
@@ -0,0 +1,383 @@
+            d2h_stream
+                ->ThenMemcpy(static_cast<void*>(data + offsets[i] / sizeof(float)),
+                             wrapped_src,
+                             static_cast<uint64>(offsets[i + 1] - offsets[i]))
+                .ok();
+        OP_REQUIRES_ASYNC(context, success,
+                          errors::Internal("D2H memcpy failed to be enqueued."), done);
+      }
+      // TODO(zongheng): does std::move() give better performance?
+      context->device()->tensorflow_gpu_device_info()->event_mgr->ThenExecute(
+          d2h_stream, std::move(wrapped_callback));
+#endif
+    }
+  }
+
+ private:
+  std::string plasma_store_socket_name_;
+  std::string plasma_manager_socket_name_;
+
+  mutex mu_;
+  bool connected_ = false;
+  plasma::PlasmaClient client_ GUARDED_BY(mu_);
+};
+
+static Stream* h2d_stream = nullptr;
+static mutex h2d_stream_mu;
+
+// Get:  plasma -> tf.Tensor.
+template <typename Device>
+class PlasmaToTensorOp : public AsyncOpKernel {
+ public:
+  explicit PlasmaToTensorOp(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("plasma_store_socket_name",
+                                             &plasma_store_socket_name_));
+    OP_REQUIRES_OK(context, context->GetAttr("plasma_manager_socket_name",
+                                             &plasma_manager_socket_name_));
+    mutex_lock lock(mu_);
+    if (!connected_) {
+      VLOG(1) << "Connecting to Plasma...";
+      ARROW_CHECK_OK(client_.Connect(plasma_store_socket_name_,
+                                     plasma_manager_socket_name_,
+                                     PLASMA_DEFAULT_RELEASE_DELAY));
+      VLOG(1) << "Connected!";
+      connected_ = true;
+    }
+  }
+
+  ~PlasmaToTensorOp() override {
+    {
+      mutex_lock lock(mu_);
+      ARROW_CHECK_OK(client_.Disconnect());
+    }
+    {
+      mutex_lock lock(h2d_stream_mu);
+      if (h2d_stream != nullptr) {
+        delete h2d_stream;
+      }
+    }
+  }
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    const Tensor& plasma_object_id = context->input(0);
+    CHECK_EQ(plasma_object_id.NumElements(), 1);
+    const string& plasma_object_id_str = plasma_object_id.flat<std::string>()(0);
+
+    VLOG(1) << "plasma_object_id_str: '" << plasma_object_id_str << "'";
+    const plasma::ObjectID object_id =
+        plasma::ObjectID::from_binary(plasma_object_id_str);
+
+    plasma::ObjectBuffer object_buffer;
+    {
+      mutex_lock lock(mu_);
+      // NOTE(zongheng): this is a blocking call.  We might want to (1) make
+      // Plasma asynchronous, (2) launch a thread / event here ourselves, or
+      // something like that...
+      ARROW_CHECK_OK(client_.Get(&object_id, /*num_objects=*/1,
+                                 /*timeout_ms=*/-1, &object_buffer));
+    }
+
+    const int64_t size_in_bytes = object_buffer.data->size();
+    TensorShape shape({size_in_bytes / sizeof(float)});
```

Comment:
We should make sure we support different dtypes and preserve the tensor type.

This code looks like it might fail for tensors whose dtype is smaller than 4 bytes.